### DISCUSSION TITLE
How can TCA works compatible with ObjC?

### DISCUSSION THREAD

**Discussion by alexlee002** [pointfreeco/swift-composable-architecture#1502] [Q&A]
I want to import TCA into my project using UIKit,  most of the code of my project is still using ObjC,  the rest are using Swift.

How can TCA works compatible with ObjC?  Thanks a lot!

**Comment by mbrandonw** [comment]
Hi @alexlee002, TCA is a pure Swift library and so pretty much none of it is accessible from Objective-C. You could make Objective-C friendly wrappers for each of your TCA features (e.g. state, store, etc.), but it would be a huge amount of boilerplate, and probably not worth it.

**Reply by alexlee002** [reply]
Yeah, I want to find out a better architecture to improve my project,   But there are still many ObjC codes, and I need to support from iOS 11,  Maybe I need to use MVVM。

**Reply by stephencelis** [reply]
It's hard to know what your exact requirements are, but if you can write _some_ Swift, then those files could house any TCA domains. You just need to expose Objective-C-friendly types for whatever Objective-C files you must maintain.

For example you could write a view controller in Swift and have it use TCA internally, but if any Objective-C code needs to interface with this view controller and feed actions into the store, you'd need to expose Objective-C-friendly methods on it that coordinates that work.

Brandon and I have gotten TCA to work in layers of Objective-C in the past, but like he said, the boilerplate did not feel worth it, but maybe for a few small touch points for legacy code it wouldn't be so bad.

**Comment by lightandshadow68** [comment]
One approach is to migrate the top-level of your application to use the new SwiftUI lifecycle, then present and orchestrate your existing objective-c / UIKit content by wrapping the top level view controller in UIViewcontrollerRepresentable. From there, you can incrementally update existing Swift features to use TCA state, start creating new features with SwiftUI and TCA, etc. 

For example, in one app, we had a top-level Obj-C custom view controller container that presented each main section's UI. We wrapped that in UIViewcontrollerRepresentable to provided a declarative interface for switching sections, etc. New features had child states that were pulled back from the new top-level state. 

Eventually, you can replace existing Obj-C sections of your app to complete the process. You might need to do this sooner, than later, if you need to show Obj-C / UIKit views in SwiftUI TCA sections of your app.
